### DISCUSSION TITLE
Advanced smart form component and validation with resolver not working

### DISCUSSION THREAD

**Discussion by smeads-sweater** [react-hook-form/react-hook-form#6968] [Q&A]
I'm currently working on a React Native project, specifically migrating from formik to react-hook-form. It has been a positive experience so far, but I have run into an issue with validation. 

I've created a `Form` component based on react-hook-form's advanced [Smart Form Component](https://react-hook-form.com/advanced-usage#SmartFormComponent) example. I've also created a custom `TextInput` that will forward the input ref. This setup allows for auto focusing the "next" input after pressing "return" on the IOS/Android keyboard (this is working as expected).

For validation, I'm using `yupResolver` and passing in the validation schema on the `useForm` `resolver` param. I'm also setting the `mode` to `onSubmit`. I only want to display validation errors when the user submits the form. However, when I use the `Yup.string().email().required()` chained methods in the validation schema the validation strategy (mode) is ignored and the first character entered by the user fires the validation with errors.  If I just use `Yup.string().required()` everything works as expected; validations fire on submit.

**Validation Schema** (using .email())
```
const schema = Yup.object().shape({
  email: Yup.string()
    .email('Please provide a valid email.')
    .required('Please provide your email.'),
});
```

I'm not sure if the validation schema has anything to do with the validation ignoring the mode "onSubmit". Might be a red haring? 

Another weird issue I'm experiencing is in the `Form` component. The `onChangeText` prop calls `setValue` (as seen below). If I remove `shouldValidate` from the config object (3rd argument passed into setValue) the input does not update the value.

**Form Component** (removing shouldValidate breaks the input)
```
onChangeText: (value: string) =>
  setValue(child.props.name, value, { shouldValidate: true }),
```

Here is a [link](https://snack.expo.dev/@smeads/react-hook-form) to the implementation on snack.expo. All scenarios above are reproducible in the sandbox env.

I've got a feeling my issue is with either the `Form` component, `Yup`, or `yupResolver`.  Any suggestions would be greatly appreciated.

**Comment by bluebill1049** [comment]
you probably should use `Controller` instead of doing that manual ' setValue`.

https://react-hook-form.com/get-started#ReactNative
https://react-hook-form.com/api/usecontroller/controller
https://react-hook-form.com/api/usecontroller

**Reply by smeads-sweater** [reply]
Do you have an example of how that would work with the "Smart Form Component"? The reason I'm using the "Smart Form Component" approach is to manage auto focusing the next input when the user presses the keyboard's "return". Is there still a way to do this using `Controller`?

As seen here https://snack.expo.dev/@smeads/react-hook-form
